<h1>Linear Regression</h1>
<h2>Data 602: Week04 Assignment</h2>
<h3>Submitted by: Srashti Soni</h3>

In [1]:
import numpy as np
import pandas as pd 
from sklearn import datasets
from sklearn import linear_model

<h3>Simple Regression Using statsmodel</h3>

In [2]:
boston = datasets.load_boston()

C:\Users\srash\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this case special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing data

In [3]:
print(boston.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

In [4]:
df = pd.DataFrame(boston.data, columns=boston.feature_names)

In [5]:
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [6]:
df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT  
count  506.000000  
mean    12.653063  
std      7.141062  
min      1.730000  
25%      6.950000  
50%     11.360000  
75%     16.955000  
max     37.970000

In [7]:
#Assigning a different dataframe for housing value-'MEDV' and printing the first 5 rows
target = pd.DataFrame(boston.target, columns=["MEDV"])

In [8]:
target.head()

MEDV
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2

In [9]:
import statsmodels.api as sm
#Define 'RM' as X or the feature and 'MEDV' as y or the label
X = df["RM"]
y = target["MEDV"]


result = sm.OLS(y, X).fit()
prediction = result.predict(X) 

# Printing the statistics
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   MEDV   R-squared (uncentered):                   0.901
Model:                            OLS   Adj. R-squared (uncentered):              0.901
Method:                 Least Squares   F-statistic:                              4615.
Date:                Sun, 27 Feb 2022   Prob (F-statistic):                   3.74e-256
Time:                        16:36:08   Log-Likelihood:                         -1747.1
No. Observations:                 506   AIC:                                      3496.
Df Residuals:                     505   BIC:                                      3500.
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
RM             3.6534      0.054     67.930      0.000       3.548       3.759
==============================================================================
Omnibus:                       83.295   Durbin-Watson:                   0.493
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              152.507
Skew:                           0.955   Prob(JB):                     7.65e-34
Kurtosis:                       4.894   Cond. No.                         1.00
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<h3>Multiple Regression Using statsmodel </h3>                                    

In [10]:
X = df[['RM', 'LSTAT']]
y = target['MEDV']
model = sm.OLS(y, X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                   MEDV   R-squared (uncentered):                   0.948
Model:                            OLS   Adj. R-squared (uncentered):              0.948
Method:                 Least Squares   F-statistic:                              4637.
Date:                Sun, 27 Feb 2022   Prob (F-statistic):                        0.00
Time:                        16:36:08   Log-Likelihood:                         -1582.9
No. Observations:                 506   AIC:                                      3170.
Df Residuals:                     504   BIC:                                      3178.
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
RM             4.9069      0.070     69.906      0.000       4.769       5.045
LSTAT         -0.6557      0.031    -21.458      0.000      -0.716      -0.596
==============================================================================
Omnibus:                      145.153   Durbin-Watson:                   0.834
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              442.157
Skew:                           1.351   Prob(JB):                     9.70e-97
Kurtosis:                       6.698   Cond. No.                         4.72
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

<h3>What does a coefficient of determination of R2= .948 mean? </h3>

 R squared value of 0.948 means that the model explains 94.8% of variance in the dependent variable.

<h3>Is there any autocorrelation? If not, why? </h3>

According to durbin Watson Test, if a autocorrelation is in between 0 and 2 the autocorrelation is positive. 
In this case, the durbin watson test value is 0.834 which is in between 0 and 2.

<h3>Are the data coming from a normal distribution? Why? </h3>

Through the inference of OLS we got the skew value(1.351) which is positive, that means the distribution is right skewed and not normally distributed.

<h3>Multiple Regression Using scikitlearn</h3>

In [11]:
X = df[['RM', 'LSTAT']]
y = target['MEDV']

In [12]:
lm = linear_model.LinearRegression()
model = lm.fit(X,y)

In [13]:
predictions = lm.predict(X)
print(predictions)

[28.94101368 25.48420566 32.65907477 32.40652    31.63040699 28.05452701
 21.28707846 17.78559653  8.10469338 18.24650673 17.99496223 20.73221309
 18.5534842  23.64474107 23.10895823 22.9239452  24.65257604 19.73611045
 18.9297215  20.57377596 13.51732408 20.14832175 17.90896697 15.48764606
 18.35281036 16.56210901 18.74440281 18.34995811 23.51018847 24.94888935
 13.23095259 21.20092715 11.15596625 15.89983805 16.63398622 22.65107562
 21.07107521 22.81275431 22.53014238 29.46686594 33.15564849 30.0244275
 26.33937234 25.50630935 23.42747337 21.03183392 19.03080004 17.28696205
  6.35742724 16.77652446 20.38222834 23.73891662 28.42223975 23.78518476
 19.13293549 32.4841017  27.4553513  30.83048667 25.54262118 22.91599173
 19.44389291 19.76157796 27.21060683 26.99027936 29.66411644 27.68813019
 21.54751591 23.38578845 18.73350058 22.97822472 27.01833368 22.66525802
 25.99579831 25.61529631 26.24614271 24.92488095 22.94287168 23.32670532
 22.46574406 22.72305097 29.51629037 27.72630168 26.

In [14]:
#R^2 value
lm.score(X,y)

0.6385616062603403

In [15]:
#printing the coefficients
lm.coef_

array([ 5.09478798, -0.64235833])

In [16]:
cdf = pd.DataFrame(lm.coef_,X.columns,columns=['Coeff'])
cdf

Coeff
RM     5.094788
LSTAT -0.642358

In [17]:
#intercept
print(lm.intercept_)

-1.3582728118745173
